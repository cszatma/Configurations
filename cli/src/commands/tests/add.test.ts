import { spawn } from 'child_process';
import fs from 'fs-extra';
import 'jest';
import path from 'path';
import configTypes from '../../utils/config-types';

const CONFIG_GEN_PATH = require.resolve('../../config-gen');
const TS_NODE = require.resolve('ts-node/dist/bin');

const resolveFile = (file: string) => path.resolve(__dirname, file);

describe('Tests for the add command', () => {
    beforeAll(() => {
        fs.writeFileSync(resolveFile('package.json'), JSON.stringify({}));
    });

    afterAll(() => {
        ['.prettierrc.json', '.eslintrc.js', 'package.json'].forEach(file =>
            fs.unlinkSync(resolveFile(file)),
        );
    });

    it('should print the help for config-gen', () => {
        return runConfigGenCLI('--help').then(({ stdout }) => {
            expect(stdout).toMatchSnapshot();
        });
    });

    it('should print the help for config-gen add', () => {
        return runConfigGenCLI('add --help').then(({ stdout }) => {
            expect(stdout).toMatchSnapshot();
        });
    });

    it('should create a .prettierrc.json file', () => {
        return runCLIAndAssertFileOutput('add prettier', '.prettierrc.json');
    });

    it('should create a .eslintrc.js file', () => {
        return runCLIAndAssertFileOutput(
            'add .eslintrc --type js',
            '.eslintrc.js',
        );
    });

    it('should return an error since eslintrc.js already exists', () => {
        return runCLIAndAssertError('add eslint --type js');
    });

    it('should create the file even though it exists since -w is passed', () => {
        return runCLIAndAssertFileOutput('add prettier -w', '.prettierrc.json');
    });

    it('should add an eslintConfig object to package.json', () => {
        return runCLIAndAssertPackageJson('add eslint -P', 'eslint');
    });
});

interface CLIStatus {
    stdout?: string;
    stderr?: string;
    code: number;
}

function runConfigGenCLI(args: string): Promise<CLIStatus> {
    return new Promise<CLIStatus>((resolve, reject) => {
        let stdout = '';
        let stderr = '';

        const childProcess = spawn(TS_NODE, [CONFIG_GEN_PATH, args], {
            cwd: __dirname,
            shell: true,
        });

        childProcess.on('error', error => reject(error));
        childProcess.stdout.on('data', data => {
            stdout += data.toString();
        });

        childProcess.stderr.on('data', data => {
            stderr += data.toString();
        });

        childProcess.on('close', code => {
            if (stderr) {
                reject({ stderr, code });
            } else {
                resolve({ stdout, code });
            }
        });
    });
}

function runCLIAndAssertFileOutput(args: string, fileName: string) {
    return runConfigGenCLI(args).then(({ stdout, code }) => {
        expect(stdout).toMatchSnapshot();
        expect(code).toBe(0);
        expect(fs.existsSync(resolveFile(fileName))).toBe(true);
    });
}

function runCLIAndAssertPackageJson(args: string, configName: string) {
    return runConfigGenCLI(args).then(({ stdout, code }) => {
        expect(stdout).toMatchSnapshot();
        expect(code).toBe(0);

        const packageJson = require(resolveFile('package.json'));
        const template = require(resolveFile('../../templates/' + configName));
        const configPackageJsonName =
            configTypes[configName].fileNames.packageJson;
        expect(packageJson[configPackageJsonName]).toEqual(template);
    });
}

function runCLIAndAssertError(args: string) {
    return runConfigGenCLI(args).catch(({ stderr, code }) => {
        expect(stderr).toMatchSnapshot();
        expect(code).toBe(1);
    });
}
