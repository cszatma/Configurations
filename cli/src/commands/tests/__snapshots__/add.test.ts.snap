// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for the add command should add an eslintConfig object to package.json 1`] = `
"Successfully wrote eslint config to package.json.
"
`;

exports[`Tests for the add command should create a .eslintrc.js file 1`] = `
"Successfully wrote eslint config to /Users/christopherszatmary/Developer/Global/Configurations/cli/src/commands/tests/.eslintrc.js.
"
`;

exports[`Tests for the add command should create a .prettierrc.json file 1`] = `
"Successfully wrote prettier config to /Users/christopherszatmary/Developer/Global/Configurations/cli/src/commands/tests/.prettierrc.json.
"
`;

exports[`Tests for the add command should create the file even though it exists since -w is passed 1`] = `
"Successfully wrote prettier config to /Users/christopherszatmary/Developer/Global/Configurations/cli/src/commands/tests/.prettierrc.json.
"
`;

exports[`Tests for the add command should print the help for config-gen 1`] = `
"
  Usage: config-gen [options] [command]

  Options:

    -v, --version                output the version number
    -h, --help                   output usage information

  Commands:

    add [options] <config-name>  Adds the given configuration file to your project
    list|ls [config]             List all available configs or list available file types for a given config
    help

Run \`config-gen COMMAND --help\` for more information about a given command.
"
`;

exports[`Tests for the add command should print the help for config-gen add 1`] = `
"
  Usage: add <config-name> [options]

  Adds the given configuration file to your project

  Options:

    -p, --path <path>         Path to add config file, defaults to current directory
    -c, --create-directories  Automatically creates intermediate directories in the given path if they do not exist
    -t, --type <type>         File type, either js or json, defaults to json
    -i, --indent <indent>     Sets the indentation amount for the config file, defaults to 2
    -P, --package-json        Adds the config file to your package.json
    -w, --write               Overwrites a config file if it already exists
    -f, --force               Equivalent to calling both --create-directories and --write
    -h, --help                output usage information
"
`;

exports[`Tests for the add command should return an error since eslintrc.js already exists 1`] = `
".eslintrc.js already exists! Aborting.
"
`;
